server/src/main/java/org/spine3/server/QueryService.java

```diff
@@ -0,0 +1,143 @@
+/*
+ *
+ * Copyright 2016, TeamDev Ltd. All rights reserved.
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+package org.spine3.server;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+import io.grpc.stub.StreamObserver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.spine3.client.Query;
+import org.spine3.client.QueryResponse;
+import org.spine3.client.grpc.QueryServiceGrpc;
+import org.spine3.protobuf.KnownTypes;
+import org.spine3.protobuf.TypeUrl;
+
+import java.util.Set;
+
+/**
+ * The {@code QueryService} provides a synchronous way to fetch read-side state from the server.
+ *
+ * <p> For asynchronous read-side updates please see {@link SubscriptionService}.
+ *
+ * @author Alex Tymchenko
+ */
+public class QueryService extends QueryServiceGrpc.QueryServiceImplBase {
+
+    private final ImmutableMap<TypeUrl, BoundedContext> typeToContextMap;
+
+    private QueryService(Builder builder) {
+        this.typeToContextMap = builder.getBoundedContextMap();
+    }
+
+    public static Builder newBuilder() {
+        return new Builder();
+    }
+
+    @SuppressWarnings("RefusedBequest") // as we override default implementation with `unimplemented` status.
+    @Override
+    public void read(Query query, StreamObserver<QueryResponse> responseObserver) {
+
+        log().debug("Incoming query: {}", query);
+        final String typeAsString = query.getTarget()
+                                         .getType();
+
+        final TypeUrl type = KnownTypes.getTypeUrl(typeAsString);
+        final BoundedContext boundedContext = typeToContextMap.get(type);
+
+        try {
+            boundedContext.getStand()
+                          .execute(query, responseObserver);
+
+        } catch (@SuppressWarnings("OverlyBroadCatchBlock") Exception e) {
+            log().error("Error processing query", e);
+            responseObserver.onError(e);
+            responseObserver.onCompleted();
```

Comment:
The Javadoc of `StreamObserver.onError` says:
> May only be called once and if called it must be the last method called.

According to this, I believe we should not call `onCompleted()`.